[simple theme] SearXNG wordmark & favicon

return42 opened the pull request:
## What does this PR do?

[simple theme] SearXNG wordmark & favicon

## Why is this change important?

["the favicon is not really suited for dark mode (basically we need different icons for dark and for light mode eventually)"](https://github.com/searxng/searxng/pull/430#issuecomment-951967239)

## Author's checklist

Favicon (see tab) and wordmark (left to the search form)  has been changed ..

![grafik](https://user-images.githubusercontent.com/554536/140060455-15c69544-bfc4-49e1-a391-10a24d55dba0.png)

before this PR:

![image](https://user-images.githubusercontent.com/1594191/138556054-120acfa0-887d-4f86-b195-55762ffabfe8.png)

## Related issues

https://github.com/searxng/searxng/pull/378

dalf commented on searx/templates/simple/search.html:
```diff
@@ -1,6 +1,6 @@
 <form id="search" method="{{ method or 'POST' }}" action="{{ url_for('search') }}">
   <a id="search_logo" href="{{ url_for('index') }}">
-    <img class="search_logo_img" src="{{ url_for('static', filename='img/favicon.png') }}">
+    <img class="search_logo_img" src="{{ url_for('static', filename='img/searxng-wordmark.svg') }}">
```
To avoid an additional HTTP request, it can be replaced by an optimized inline version of the SVG file (429 bytes) :
```svg
<svg viewBox="0 0 92 92" version="1.1"><g transform="translate(-40.92 -17.42)"><circle cx="75.92" cy="53.9" r="30" fill="none" stroke="#3050ff" stroke-width="10"/><path d="M67.51 37.92a18 18 0 0 1 21.06 3.3 18 18 0 0 1 3.13 21.09" fill="none" stroke="#3050ff" stroke-width="5"/><rect width="18.85" height="39.96" x="3.71" y="122.09" ry="0" transform="rotate(-46.23)" fill="#3050ff"/></g></svg>
```

This version can be an external file that is included using the Jinja2 syntax.

The version above was made using https://jakearchibald.github.io/svgomg/ with these options:
* set precision to 2
* check "prefer viewbox to width/height" (at the bottom of the options)
* Manually remove `xmlns="http://www.w3.org/2000/svg"`.

For reference:
* https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/SVG_In_HTML_Introduction
* how the GitHub logo in the top left corner in embedded.

---

[EDIT] Also, the PNG file size can be optimized and reduced to 1.2KB.

return42 replied:
> This version can be an external file that is included using the Jinja2 syntax.

I assume you are talking about something like this ... ?
```python

from flask import Markup
---
    return render(
        'results.html',
        wordmark_svg = Markup(open('.....img/searxng-wordmark.svg').read)
```
> To avoid an additional HTTP request

Yes, we can avoid one request, but on the opposite (if I'm not wrong) we lost flexibility with the `static/themes/simple` folder 

https://github.com/searxng/searxng/blob/580a55fe866b82e93dec6bc4e07fa8a0f2056475/searx/settings.yml#L64-L66

dalf replied:
What about 

webapp.py:
```python
wordmark_contents = {}
for indice, theme in enumerate(themes):
    ...
    workmark_file = os.path.join(settings['ui']['static_path'], 'theme', theme, 'img/searxng-wordmark.svg')
    if os.path.isfile(workmark_file):
        with os.open(workmark_file, 'rb') as f:
            wordmark_contents[theme] = f.read()
    ...


def render(...):
  ...
  kwargs['wordmark_content'] = wordmark_contents.get(kwargs['theme'])
  ...
```

search.html:
```jinja2
{{ wordmark_content }}
```

?

[EDIT] Too complicated.

Perhaps this:
* add a new function : `content_for`, same as `url_for` but actually returns the content.
* `kwargs['content_for'] = content_for` 

--> `{{ content_for('img/searxng-wordmark.svg') }}`

return42 replied:
@dalf I have tried what you suggested, see dd4da6f / but now I have a problem to size the inline SVG, see 
https://github.com/searxng/searxng/pull/473/commits/dd4da6f465349488ef101d76773d630dffea0d24#r743818450

return42 commented on searx/templates/simple/search.html:
```diff
@@ -1,6 +1,8 @@
-    <img class="search_logo_img" src="{{ url_for('static', filename='img/searxng-wordmark.svg') }}">
+    <div class="search_logo_img">
+      {{ insert_xml('img/searxng-wordmark.svg') }}
+    </div>
```
this leads into  ..

![grafik](https://user-images.githubusercontent.com/554536/140546001-fabc20b7-e5ef-4867-8bce-7eca6fe39608.png)

 and I do not know how I can set a fixed size for the `<div>` container or how the can set a fixed box-size for the  _inline SVG_ ..

I'm not an CSS expert and I haven't understand the CSS box model really.

return42 committed:
[simple theme] SearXNG wordmark (SVG)

Related-to:

- https://github.com/searxng/searxng/pull/430#issuecomment-951967239
- https://github.com/searxng/searxng/pull/378

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[enh] minify searxng-wordmark.svg (HTMLMinifier)

Command::

  ./node_modules/.bin/html-minifier \
      --remove-comments \
      --collapse-whitespace \
      src/svg/searxng-wordmark.svg \
      -o ../../../templates/__common__/searxng-wordmark.min.svg

- html-minifier: https://github.com/kangax/html-minifier
  - onilne: https://kangax.github.io/html-minifier
  - grunt: https://www.npmjs.com/package/grunt-contrib-htmlmin
  - grunt-contrib-htmlmin: https://github.com/gruntjs/grunt-contrib-htmlmin
  - npm: https://www.npmjs.com/package/html-minifier

To test, rebuild your node environment::

  make node.env

Alternatives:

- pretty-data: https://github.com/vkiryukhin/pretty-data
  - grunt: https://www.npmjs.com/package/grunt-xmlmin
  - grunt-xmlming: https://github.com/dtrunk90/grunt-xmlmin
  - npm: https://www.npmjs.com/package/grunt-xmlmin

- minify-xml: https://github.com/kristian/minify-xml
  - no grunt package available
  - npm: https://www.npmjs.com/package/minify-xml

src/svg/searxng-wordmark.svg':
'../../../templates/__common__/searxng-wordmark.min.svg'

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[simple theme] create favicon.png from src/svg/searxng-wordmark.svg

new bash function convert_if_newer() / usage::

    convert_if_newer <origfile> <outfile> [<options>, ...]
    convert_if_newer "path/to/origin.svg" "path/to/converted.png" -transparent white -resize 64x64

Run's ImageMagik' convert comand to generate <outfile> from <origfile>, if
<origfile> is newer than <outfile>.  The command line is to convert is::

    convert <origfile> [<options>, ...] <outfile>

PNG 'searx/static/themes/simple/img/favicon.png' has been created by::

  $ make themes.simple
  CONVERT   searx/static/themes/simple/src/svg/searxng-wordmark.svg -transparent white -resize 64x64 searx/static/themes/simple/img/favicon.png
  ...

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[simple theme] replace old searx logo by searxng-wordmark.min.svg

Suggested-by: @dalf https://github.com/searxng/searxng/pull/473#issuecomment-962476683
Suggested-by: @MrPaulBlack https://github.com/searxng/searxng/pull/473#issuecomment-962579826
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[simple theme] SearXNG wordmark (plain SVG)

Remove XML namespaces from Incscape [1]::

   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"

[1] https://wiki.inkscape.org/wiki/PlainSVG

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

mrpaulblack reviewed (commented):
:-1: The favicon.png does not have a alpha channel...
![image](https://user-images.githubusercontent.com/38733246/140829675-c9c53e1f-5ebc-4fc8-bdf9-3d8fd02ebd27.png)

Is this intended? (Or to rephrase: is it possible to generate favicon.png with an alpha channel from the svg?)

Everything else LGTM:
:+1:  the `make themes.all` and `make themes.simple` exec the htmlmin step for simple properly
:+1: watermark displays correctly on desktop, tablet and mobile
:+1: the jinja2 template inlines the svg correctly into the html

dalf commented on searx/static/themes/simple/package.json:
```diff
@@ -8,6 +8,8 @@
     "grunt-contrib-less": "~3.0.0",
     "grunt-contrib-uglify": "~5.0.1",
+    "grunt-xmlmin": "~0.1.8",
+    "grunt-contrib-htmlmin": "~3.1.0",
```
[grunt-image](https://www.npmjs.com/package/grunt-image) uses [svgo](https://www.npmjs.com/package/svgo). The min.svg file size can be reduced by ~ half (it removes the sodipodi elements for example).

return42 replied:
OK I understand .. can we optimize this later?  First I think we should clarify which version of svgo we should use. ATM we have an issue reported when `make node.clean node.env` is build:

    npm WARN deprecated svgo@0.6.6: This SVGO version is no longer supported. Upgrade to v2.x.x.

This issue is caused by grunt-webfont and might be also related to https://github.com/searxng/searxng/pull/488#issuecomment-963606592 / Hint: https://github.com/sapegin/grunt-webfont is not maintained since years, see archived repository at https://github.com/sapegin/grunt-webfont and we have other issues with, see  https://github.com/searxng/searxng/pull/248#issuecomment-894672328

return42 commented on manage:
```diff
@@ -692,11 +692,44 @@ themes.oscar() {
+    local src="searx/static/themes/simple/src"
+    local static="searx/static/themes/simple"
+    (   set -e
+        convert_if_newer "$src/svg/searxng-wordmark.svg" "$static/img/favicon.png" -background none -resize 64x64
```
> :-1: The favicon.png does not have a alpha channel... ![image](https://user-images.githubusercontent.com/38733246/140829675-c9c53e1f-5ebc-4fc8-bdf9-3d8fd02ebd27.png) Is this intended? (Or to rephrase: is it possible to generate favicon.png with an alpha channel from the svg?)

Oops .. it is fixed right here by adding [`-background none`](https://www.imagemagick.org/script/command-line-options.php#background) to the convert command.

dalf replied:
I wish to keep all the build steps inside gruntfiles.js

In the simple theme directory, `npm run watch` allows to rebuild the theme a files change. If we mix bash and node it starts to be complicated to do that.

There are some grunt modules to convert to svg to png, but I haven't tried them.

Otherwise it is possible to call a command line from grunt: https://www.npmjs.com/package/grunt-exec

But, the logo won't change as often the other files, so IMO, it is not an issue if there is automatic build of the favicon.

return42 replied:
> I wish to keep all the build steps inside gruntfiles.js

Hm, .. I understand to want to have all theme build-tasks at one place (grunt), but the drawback I see is; with [grunt-exec](https://www.npmjs.com/package/grunt-exec) we can't express a conditional _convert if origin is newer_. IMO calling shell tasks from grunt increase complexity unneeded.

return42 commented on manage:
```diff
@@ -692,11 +692,45 @@ themes.oscar() {
+    local src="searx/static/themes/simple/src"
+    local static="searx/static/themes/simple"
+    (   set -e
+        convert_if_newer "$src/svg/searxng-wordmark.svg" "$static/img/favicon.png" \
+                         -transparent white -resize 64x64
```
> The favicon.png does not have an alpha channel:

@dalf thanks a lot for your attention! ... I'm very sorry for the mess .. it is not ImageMagick what is broken .. it is me 🤦🏻 

Its now fixed ...

![grafik](https://user-images.githubusercontent.com/554536/141513942-ec987d71-6286-4e3a-bc90-4905b12b2b9a.png)

mrpaulblack replied:
LGTM :+1: 
![image](https://user-images.githubusercontent.com/38733246/141597424-592a48f1-4ab1-4afd-97b4-5f55b81155df.png)

return42 committed:
[build] /static
